lib/spack/spack/filesystem_view.py

```diff
@@ -0,0 +1,494 @@
+        standalones = specs - extensions
+
+        set(map(self._check_no_ext_conflicts, extensions))
+        # fail on first error, otherwise link extensions as well
+        if all(map(self.activate_standalone, standalones)):
+            all(map(self.activate_extension, extensions))
+
+    @with_extensions
+    def activate_extension(self, spec):
+        if not spec.package.is_extension:
+            tty.error(self._croot + 'Package %s is not an extension.'
+                      % spec.name)
+            return False
+
+        try:
+            if not spec.package.activated:
+                spec.package.do_activate(verbose=self.verbose)
+
+        except ExtensionAlreadyInstalledError:
+            # As we use sets in activate(), the order in which packages get
+            # activated is essentially random. So this spec might have already
+            # been activated as dependency of another package -> fail silently
+            pass
+
+        # make sure the meta folder is linked as well (this is not done by the
+        # extension-activation mechnism)
+        if not self.check_active(spec):
+            self.link_meta_folder(spec)
+
+        return True
+
+    def activate_standalone(self, spec):
+        if self.check_active(spec):
+            tty.warn(self._croot + 'Skipping already linked package: %s'
+                     % spec.cshort_spec)
+            return True
+
+        tree = LinkTree(spec.prefix)
+
+        if not self.ignore_conflicts:
+            conflict = tree.find_conflict(self.root)
+            if conflict is not None:
+                tty.error(self._croot +
+                          "Cannot link package %s, file already exists: %s"
+                          % (spec.name, conflict))
+                return False
+
+        conflicts = tree.merge(self.root, link=self.link,
+                               ignore=ignore_metadata_dir,
+                               report_existing=self.ignore_conflicts)
+        self.link_meta_folder(spec)
+
+        if self.ignore_conflicts:
+            for c in conflicts:
+                tty.warn(self._croot + "Could not link: %s" % c)
+
+        if self.verbose:
+            tty.info(self._croot + 'Linked package: %s' % spec.cshort_spec)
+        return True
+
+    def check_active(self, spec):
+        assert spec.concrete
+        return spec == self.get_spec(spec)
+
+    def deactivate(self, *specs, **kwargs):
+        assert all((s.concrete for s in specs))
+
+        if kwargs.get("with_dependencies", False):
+            specs = get_dependencies(specs)
+
+        if kwargs.get("exclude", None):
+            specs = set(filter_exclude(specs, kwargs["exclude"]))
+
+        to_deactivate = set(specs)
+
+        all_specs = set(self.get_all_specs())
+        to_keep = all_specs - to_deactivate
+
+        # remove all packages depending on the ones to remove
+        to_deactivate.update(find_dependents(to_keep, to_deactivate))
```

Comment:
Got it. I would prefer there would be an option for this like ```--remove-dependents```. If it is true by default, warn the user when you are automatically removing dependents; if it is false by default, warn the user when you are leaving "detached" dependents around.